### PR TITLE
Feature/separate classifier

### PR DESCRIPTION
**Pull request by LoyVanBeek**
Classifier is separated into a separate class. 

TODO: 

- [x] Do not save to file before passing to tensorflow session. TF has a function to convert a numpy array to a tensor, but at what tensor to substitute it in the feed_dict?
- [x] TensorFlow can apparently only run when sess.run is called from the main thread. The synchronisation code to deal with this is ugly (but works!)
- [x] To make the ObjectRecognizer generic and not specific to a particular graph: allow to specify the name of the input and output tensors.

### CODE REVIEW

**Comment by reinzor** [comment]
Please merge with current master then :)

**Comment by LoyVanBeek** [comment]
I rebased the branch on master. This fixes the build

**Comment by LoyVanBeek** [comment]
The ugly threading and syncing might not be needed anymore: 
https://stackoverflow.com/questions/38694111/is-it-thread-safe-when-using-tf-session-in-inference-service

But maybe: https://www.tensorflow.org/versions/r0.12/how_tos/threading_and_queues/ (but be sure to check for r1.3 wre' currently on)

File: image_recognition_util/src/image_recognition_util/image_reader.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import fnmatch
 
 
-def read_annotated(dir_path, patterns=["*.jpg", "*.png", "*.jpeg"]):
+def read_annotated(dir_path, patterns=["*.jpg", "*.png", "*.jpeg"], include_path=False):
```
Comment (reinzor) [inline image_recognition_util/src/image_recognition_util/image_reader.py]: is include_path=False ever used? changing the return type base on input argument is kinda confusing.

Comment (reinzor) [reply image_recognition_util/src/image_recognition_util/image_reader.py]: So the code would crash if input arguments are changing :p

Comment (LoyVanBeek) [reply image_recognition_util/src/image_recognition_util/image_reader.py]: I know, the argument is there only for when you're testing. I found out that the reader kept repeating the same label over and over, also for paths of a different label. 

I'll remove the flag later on in this WIP.

**Comment by LoyVanBeek** [comment]
@reinzor can you check this PR again?

The script 'evaluate_classifier' works but is not fully done. It's not really part of this PR and can be improved still. This is a preliminary version, maybe it should be a separate PR as well.

**Comment by LoyVanBeek** [comment]
Example output of the classifier_metrics script. 
results

File: image_recognition_util/src/image_recognition_util/image_reader.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import os
+import fnmatch
+
+
+def read_annotated(dir_path, patterns=["*.jpg", "*.png", "*.jpeg"]):
+    """
+    Read annotated images from a directory. This reader assumes that the images in this directory are separated in
+    different directories with the label name as directory name. The method returns a generator of the label (string)
+    and the opencv image.
+    :param dir_path: The base directory we are going to write read
```
Comment (reinzor) [inline image_recognition_util/src/image_recognition_util/image_reader.py]: 'write read' should be 'read'

File: image_recognition_util/src/image_recognition_util/image_reader.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    :param patterns: Patterns of the images the reader should match
+    """
+    for label in os.listdir(dir_path):
+        for root, dirs, files in os.walk(os.path.join(dir_path, label)):
+            for basename in files:
+                for pattern in patterns:
+                    if fnmatch.fnmatch(basename, pattern):
+                        yield label, cv2.imread(os.path.join(root, basename))
+
+if __name__ == "__main__":
```
Comment (reinzor) [inline image_recognition_util/src/image_recognition_util/image_reader.py]: Remove this testing functionality, if you would like to test this, add a unit test in the test subdir of the package and include in the cmakelist

File: tensorflow_ros/scripts/classifier_metrics [hunk 4]
```diff
@@ -0,0 +1,315 @@
+#!/usr/bin/env python
```
Comment (reinzor) [inline tensorflow_ros/scripts/classifier_metrics]: Make sure all deps can be installed with use of `rosdep`. Also make sure that a person understands what is happening here. Is this tooling btw specific to `tensorflow_ros`?

File: tensorflow_ros/scripts/evaluate_classifier [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    labels_per_image += [label]
+    predictions_per_image += [image_predictions]
+
+    match = label == image_predictions[0][0]
+    matches += [match]
+    pprint.pprint("Image is a {} and classified as a {} ({:.3f}) --> {}".format(label, image_predictions[0][0],
+                                                                                image_predictions[0][1], match))
+    print "Matches: {} / {}".format(matches.count(True), len(matches))
+
+    # TODO: Make a np.array of the scores for each label, in a predefined order, the same for all images
```
Comment (reinzor) [inline tensorflow_ros/scripts/evaluate_classifier]: TODO?

File: tensorflow_ros/scripts/get_object_recognition [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import argparse
+from tensorflow_ros.object_recognizer import ObjectRecognizer
+import cv2
+import numpy as np
+
+import os
+import pprint
+
+# Assign description to the help doc
+parser = argparse.ArgumentParser(description='Get image classifications via some TensorFlow network')
```
Comment (reinzor) [inline tensorflow_ros/scripts/get_object_recognition]: Example usage would be nice when calling --help

File: image_recognition_util/src/image_recognition_util/image_reader.py (hunk 3 above)
Comment (LoyVanBeek) [reply image_recognition_util/src/image_recognition_util/image_reader.py]: The test is dependent on the filesystem it runs on, so no tests unfortunately.

File: tensorflow_ros/scripts/classifier_metrics (hunk 4 above)
Comment (LoyVanBeek) [reply tensorflow_ros/scripts/classifier_metrics]: It shouldn't be, should be general to basically any (multi-label) classifier actually. What's a good place to put it?

Also: it heavily depends on matplotlib, numpy and sklearn. The image_recognition_utils are otherwise quite light, so adding i to that isn't that nice either.

Comment (LoyVanBeek) [reply tensorflow_ros/scripts/classifier_metrics]: OK, the plan for this script is: move it to a new package image_recognition_tools. Because that is not part of this PR, delete it on this branch but add it (back again) later in a new branch (feature/classifier_metrics) off this PR's branch
